File: NIP/meta_measures_example.R
# Clear workspace
rm(list = ls())

# Load meta-measures

# install.packages("devtools")
# devtools::install_github("craddm/metaSDT")

library(metaSDT)
source("meta_measures_library.R")

# Example table
tab <- matrix(scan(text = "
    16 17 14 03
    02 11 19 18
"), nrow = 2, byrow = TRUE)

# Evaluate meta-d' -------------------------------------------------------------
fit_meta_d_MLE(tab[1,], tab[2,])

estimate_sensitivity(tab)
estimate_meta_sensitivity(tab)
estimate_M_ratio(tab)

# Evaluate information-theoretic measures --------------------------------------
estimate_meta_I(tab)
estimate_meta_Ir1(tab)
estimate_meta_Ir2(tab)

estimate_RMI(tab)
estimate_RMI_with_bias_reduction(tab)

# Alternative meta-d'fits  -----------------------------------------------------

library(statConfR)
stimulus <- rep(c(-1, +1), rowSums(tab))
rating <- c(rep(1:4, tab[1,]), rep(1:4, tab[2,]))
correct <- rep(c(1, 0, 0, 1), c(sum(tab[1, 1:2]), 
                                sum(tab[1, 3:4]), 
                                sum(tab[2, 1:2]), 
                                sum(tab[2, 3:4])))

dat <- data.frame(stimulus = factor(stimulus),
                  rating = factor(rating),
                  correct = correct,
                  participant = 1)

table(dat$correct, dat$rating)

fitMetaDprime(dat)

# ------------------------------------------------------------------------------
library(stats)
# install.packages("optimx")
library(optimx)
source("translated_meta_fit.R")

fit_meta_d_MLE_Meyen(tab[1,], tab[2,])

#  -----------------------------------------------------------------------------

output = system(paste('python adapt_fit_meta_d_MLE.py ', 
                paste0(tab[1, ], collapse = " "),
                paste0(tab[2, ], collapse = " ")),
                intern = TRUE)
l <- length(output)
sensitivity <- output[l-2]
meta_sensitivity <- output[l-1]
m_ratio <- output[l]

File: NIP/Python/meta_measures_example.py
import numpy as np

import os
os.chdir("C:/Users/saschameyen/Desktop/NIP_Workshop2024_Meyen/NIP/Python")
os.getcwd()

from meta_measures_library import *
from fit_meta_d_MLE import *

# Example usage:
tab = np.array([[16, 17, 14, 3], [2, 11, 19, 18]])

# Evaluate meta-d' -------------------------------------------------------------
meta_fit = fit_meta_d_MLE(tab[0,], tab[1,])

print("Estimated d'                   =", meta_fit["da"])
print("Estimated meta-d'              =", meta_fit["meta_da"])
print("Estimated M-ratio = meta-d'/d' =", meta_fit["M_ratio"])

# Evaluate information-theoretic measures --------------------------------------
print("Estimated meta-I               =", estimate_meta_I(tab))
print("Estimated meta-Ir1             =", estimate_meta_Ir1(tab))
print("Estimated meta-Ir2             =", estimate_meta_Ir2(tab))

print("Estimated RMI                  =", estimate_RMI(tab))
print("Estimated Debiased RMI         =", estimate_RMI_with_bias_reduction(tab))
